File: utilities/src/main/java/com/toocol/ssh/utilities/handler/AbstractMessageHandler.java
package com.toocol.ssh.utilities.handler;

import com.toocol.ssh.utilities.address.IAddress;
import com.toocol.ssh.utilities.execeptions.IStacktraceParser;
import com.toocol.ssh.utilities.log.Loggable;
import com.toocol.ssh.utilities.utils.Castable;
import io.vertx.core.Context;
import io.vertx.core.Vertx;
import io.vertx.core.eventbus.EventBus;
import io.vertx.core.eventbus.Message;

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/4/21 10:34
 */
public abstract class AbstractMessageHandler implements Castable, Loggable, IStacktraceParser {
    /**
     * the vertx system object.
     */
    protected final Vertx vertx;
    /**
     * the context of verticle.
     */
    protected final Context context;
    /**
     * the event bus of Vert.x
     */
    protected final EventBus eventBus;

    protected AbstractMessageHandler(Vertx vertx, Context context) {
        this.vertx = vertx;
        this.context = context;
        this.eventBus = vertx.eventBus();
    }

    /**
     * handle the message event
     *
     * @param message message event
     * @param <T>     generic type
     */
    public abstract <T> void handle(Message<T> message);

    /**
     * return the address that handler handle of.
     *
     * @return address
     */
    public abstract IAddress consume();

}

File: utilities/src/main/java/com/toocol/termio/utilities/module/AbstractMessageHandler.kt
package com.toocol.termio.utilities.module

import com.toocol.termio.utilities.execeptions.IStacktraceParser
import com.toocol.termio.utilities.log.Loggable
import com.toocol.termio.utilities.utils.Castable
import io.vertx.core.Context
import io.vertx.core.Vertx
import io.vertx.core.eventbus.EventBus
import io.vertx.core.eventbus.Message
import org.jetbrains.annotations.NotNull

/**
 * @author ZhaoZhe (joezane.cn@gmail.com)
 * @date 2022/4/21 10:34
 */
abstract class AbstractMessageHandler protected constructor(
    @JvmField
    // the vertx system object.
    protected val vertx: Vertx,
    @JvmField
    // the context of verticle.
    protected val context: Context,
) : Castable, Loggable, IStacktraceParser {

    /**
     * the event bus of Vert.x
     */
    @JvmField
    protected val eventBus: EventBus = vertx.eventBus()

    /**
     * handle the message event
     *
     * @param message message event
     * @param <T>     generic type
    </T> */
    abstract fun <T> handle(message: Message<T>)

    /**
     * return the address that handler handle of.
     *
     * @return address
     */
    @NotNull
    abstract fun consume(): IAddress
}